presto-main/src/test/java/com/facebook/presto/metadata/TestSignatureBinder.java

```diff
@@ -0,0 +1,793 @@
+                .toArguments("array(HyperLogLog)", "HyperLogLog")
+                .isFailed();
+
+        Signature fooFunction = functionSignature()
+                .returnType("T")
+                .argumentTypes("array(T)", "array(T)")
+                .typeVariableConstraints(ImmutableList.of(typeVariable("T")))
+                .build();
+
+        assertBindSignature(fooFunction)
+                .toArguments("array(bigint)", "array(bigint)")
+                .hasTypeVariablesBound(ImmutableMap.of("T", "bigint"));
+
+        assertBindSignature(fooFunction)
+                .withCoercion()
+                .toArguments("array(bigint)", "array(varchar)")
+                .isFailed();
+    }
+
+    @Test
+    public void testMap()
+            throws Exception
+    {
+        Signature getValueFunction = functionSignature()
+                .returnType("V")
+                .argumentTypes("map(K,V)", "K")
+                .typeVariableConstraints(ImmutableList.of(typeVariable("K"), typeVariable("V")))
+                .build();
+
+        assertBindSignature(getValueFunction)
+                .toArguments("map(bigint,varchar)", "bigint")
+                .hasTypeVariablesBound(ImmutableMap.of(
+                        "K", "bigint",
+                        "V", "varchar"
+                ));
+
+        assertBindSignature(getValueFunction)
+                .withCoercion()
+                .toArguments("map(bigint,varchar)", "varchar")
+                .isFailed();
+    }
+
+    @Test
+    public void testVariadic()
+            throws Exception
+    {
+        Signature mapVariadicBoundFunction = functionSignature()
+                .returnType("bigint")
+                .argumentTypes("T")
+                .typeVariableConstraints(ImmutableList.of(withVariadicBound("T", "map")))
+                .build();
+
+        assertBindSignature(mapVariadicBoundFunction)
+                .toArguments("map(bigint,bigint)")
+                .hasTypeVariablesBound(ImmutableMap.of("T", "map(bigint,bigint)"));
+
+        assertBindSignature(mapVariadicBoundFunction)
+                .toArguments("array(bigint)")
+                .isFailed();
+
+        assertBindSignature(mapVariadicBoundFunction)
+                .withCoercion()
+                .toArguments("array(bigint)")
+                .isFailed();
+
+        Signature decimalVariadicBoundFunction = functionSignature()
+                .returnType("bigint")
+                .argumentTypes("T")
+                .typeVariableConstraints(ImmutableList.of(withVariadicBound("T", "decimal")))
+                .build();
+
+        assertBindSignature(decimalVariadicBoundFunction)
+                .toArguments("decimal(2,1)")
+                .hasTypeVariablesBound(ImmutableMap.of("T", "decimal(2,1)"));
+
+        assertBindSignature(decimalVariadicBoundFunction)
+                .toArguments("bigint")
+                .isFailed();
+
+        assertBindSignature(decimalVariadicBoundFunction)
```

Comment:
For most of the test the signature is declared only once, and than several test cases are executed for that signature. So, i would rather prefer to keep signature builder and assertion in separate. I'm going to stick to 

```
assertThat(function)
    .boundTo("xxx", "yyy")
    .withCoercions()
    .produces(ImmutableMap.of(...))
```